File: SwrveSDK/SDK/Messaging/SwrveThemedUIButton.m
#import "SwrveThemedUIButton.h"
#import "SwrveSDKUtils.h"

#if __has_include(<SwrveSDKCommon/SwrveUtils.h>)
#import <SwrveSDKCommon/SwrveUtils.h>
#import <SwrveSDKCommon/SwrveLocalStorage.h>
#import <SwrveSDKCommon/SwrveLogger.h>
#else
#import "SwrveUtils.h"
#import "SwrveLocalStorage.h"
#import "SwrveLogger.h"
#endif

@interface SwrveThemedUIButton ()

@property(atomic, retain) SwrveButtonTheme *theme;
@property(atomic, retain) SwrveCalibration *calibration;
@property(atomic) CGFloat renderScale;

@end

@implementation SwrveThemedUIButton

@synthesize theme;
@synthesize calibration;
@synthesize renderScale;

- (id)initWithTheme:(SwrveButtonTheme *)buttonTheme
               text:(NSString *)text
              frame:(CGRect)frameRect
        calabration:(SwrveCalibration *)swrveCalibration
        renderScale:(CGFloat)scale {

    if (self = [super init]) {
        self.theme = buttonTheme;
        self.frame = frameRect;
        self.calibration = swrveCalibration;
        self.renderScale = scale;

        [self setTitle:text forState:UIControlStateNormal];
        [self applyTextAlignment];
        [self applyPadding];
        [self applyFont];
        [self applyFontColor];
        [self applyCornerRadius];
        [self applyBorder];
        [self applyBackground];

        self.clipsToBounds = true;
    }

    return self;
}

- (void)setHighlighted:(BOOL)highlighted {
    [super setHighlighted:highlighted];
    if (highlighted) {
        if (theme.pressedState.bgColor) {
            self.backgroundColor = [SwrveUtils processHexColorValue:theme.pressedState.bgColor];
        }
        if (theme.pressedState.borderColor) {
            UIColor *borderColor = [SwrveUtils processHexColorValue:theme.pressedState.borderColor];
            self.layer.borderColor = [borderColor CGColor];
        }
    } else {
        if (theme.bgColor) {
            self.backgroundColor = [SwrveUtils processHexColorValue:theme.bgColor];
        }
        if (theme.borderColor) {
            UIColor *borderColor = [SwrveUtils processHexColorValue:theme.borderColor];
            self.layer.borderColor = [borderColor CGColor];
        }
    }
}

- (void)applyTextAlignment {
    if (!theme.hAlign) {
        return;
    }
    if ([theme.hAlign isEqualToString:@"LEFT"]) {
        self.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
    } else if ([theme.hAlign isEqualToString:@"CENTER"]) {
        self.contentHorizontalAlignment = UIControlContentHorizontalAlignmentCenter;
    } else if ([theme.hAlign isEqualToString:@"RIGHT"]) {
        self.contentHorizontalAlignment = UIControlContentHorizontalAlignmentRight;
    }
}

- (void)applyFont {
    UIFont *defaultFont = [UIFont systemFontOfSize:theme.fontSize.floatValue];
    UIFont *titleFont = [SwrveSDKUtils fontFromFile:theme.fontFile
                                      postscriptName:theme.fontPostscriptName
                                                size:theme.fontSize.floatValue
                                               style:theme.fontNativeStyle
                                        withFallback:defaultFont];

    CGFloat scaledPointSize = theme.fontSize.floatValue;
    if (self.calibration) {
        scaledPointSize = [SwrveSDKUtils scaleFont:titleFont
                                        calibration:self.calibration
                                      swrveFontSize:theme.fontSize.floatValue
                                        renderScale:self.renderScale];
    }
    self.titleLabel.font = [titleFont fontWithSize:scaledPointSize];

    if (theme.truncate) {
        self.titleLabel.lineBreakMode = NSLineBreakByTruncatingTail;
    } else {
        self.titleLabel.numberOfLines = 1;
        self.titleLabel.lineBreakMode = NSLineBreakByWordWrapping;
        [SwrveLogger debug:@"SwrveThemedUIButton scaled size: %f", self.titleLabel.font.pointSize];
        [self scaleDownTextWithBeginSize:scaledPointSize];
        [SwrveLogger debug:@"SwrveThemedUIButton fitted size: %f", self.titleLabel.font.pointSize];
        self.titleLabel.adjustsFontSizeToFitWidth = true;
    }
}

- (void)scaleDownTextWithBeginSize:(CGFloat)fontSize {
    NSString *text = self.titleLabel.text;
    if (text.length == 0 || CGSizeEqualToSize(self.bounds.size, CGSizeZero)) {
        return;
    }

    CGFloat widthPaddingOffset = ([theme.leftPadding intValue] + [theme.rightPadding intValue] + [theme.borderWidth floatValue]) * renderScale;
    CGFloat heightPaddingOffset = ([theme.topPadding intValue] + [theme.bottomPadding intValue] + [theme.borderWidth floatValue]) * renderScale;

    CGFloat width = self.frame.size.width - widthPaddingOffset;
    CGFloat height = self.frame.size.height - heightPaddingOffset;

    CGSize constraintSize = CGSizeMake(width, height);
    while (fontSize > 1) {
        CGSize textRect = [self.titleLabel sizeThatFits:constraintSize];
        if (ceilf((float) textRect.height) <= height && ceilf((float) textRect.width) <= width) {
            break;
        }
        fontSize -= 1.0f; // Decrease the font size and try again
        self.titleLabel.font = [self.titleLabel.font fontWithSize:fontSize];
    }
}

- (void)applyFontColor {
    UIColor *titleColor = [SwrveUtils processHexColorValue:theme.fontColor];
    [self setTitleColor:titleColor forState:UIControlStateNormal];
    UIColor *titleColorPressed = [SwrveUtils processHexColorValue:theme.pressedState.fontColor];
    [self setTitleColor:titleColorPressed forState:UIControlStateHighlighted];
    if (theme.focusedState) {
        UIColor *titleColorFocused = [SwrveUtils processHexColorValue:theme.focusedState.fontColor];
        [self setTitleColor:titleColorFocused forState:UIControlStateFocused];
    }
}

- (void)applyCornerRadius {
    CGFloat cornerRadius = theme.cornerRadius.floatValue * self.renderScale;
    // check if radius is larger than half the width or half the height - if it is then use that
    if (cornerRadius >= self.frame.size.height / 2) {
        cornerRadius = self.frame.size.height / 2;
    }
    if (cornerRadius >= self.frame.size.width / 2) {
        cornerRadius = self.frame.size.width / 2;
    }
    self.layer.cornerRadius = cornerRadius;
}

- (void)applyBorder {
    if ([theme.borderWidth floatValue] > 0) {
        UIColor *borderColor = [SwrveUtils processHexColorValue:theme.borderColor];
        self.layer.borderColor = [borderColor CGColor];
        self.layer.borderWidth = [theme.borderWidth floatValue] * self.renderScale;
    }
}

- (void)applyBackground {
    if (theme.bgColor && [theme.bgColor length] > 0) {
        UIColor *bgDefault = [SwrveUtils processHexColorValue:theme.bgColor];
        self.backgroundColor = bgDefault;
    } else {
        UIImage *bgDefault = [self createUIImage:theme.bgImage];
        [self setBackgroundImage:bgDefault forState:UIControlStateNormal];
        UIImage *bgPressed = [self createUIImage:theme.pressedState.bgImage];
        [self setBackgroundImage:bgPressed forState:UIControlStateHighlighted];
        if (theme.focusedState) {
            UIImage *bgFocused = [self createUIImage:theme.focusedState.bgImage];
            [self setBackgroundImage:bgFocused forState:UIControlStateFocused];
        }
    }
}

- (UIImage *)createUIImage:(NSString *)assetName {
    NSString *cacheFolder = [SwrveLocalStorage swrveCacheFolder];
    NSString *target = [cacheFolder stringByAppendingPathComponent:assetName];
    NSURL *fileImageURL = [NSURL fileURLWithPath:target];
    NSData *fileImageData = [NSData dataWithContentsOfURL:fileImageURL];
    return [UIImage imageWithData:fileImageData];
}

- (void)applyPadding {
    CGFloat topPadding = ([theme.topPadding floatValue] + [theme.borderWidth floatValue]) * self.renderScale;
    CGFloat leftPadding = ([theme.leftPadding floatValue] + [theme.borderWidth floatValue]) * self.renderScale;
    CGFloat bottomPadding = ([theme.bottomPadding floatValue] + [theme.borderWidth floatValue]) * self.renderScale;
    CGFloat rightPadding = ([theme.rightPadding floatValue] + [theme.borderWidth floatValue]) * self.renderScale;
    self.contentEdgeInsets = UIEdgeInsetsMake(topPadding, leftPadding, bottomPadding, rightPadding);
}

@end

File: public/SwrveSDKSwift/SwrveThemedUIButton.swift
import Foundation
import UIKit

#if canImport(SwrveSDK)
import SwrveSDK
#endif

#if canImport(SwrveSDKCommon)
import SwrveSDKCommon
#endif

@objc public class SwrveThemedUIButton: SwrveUIButton {

    // MARK: - Properties

    @objc public let theme: SwrveButtonTheme
    @objc public let calibration: SwrveCalibration?
    @objc public let renderScale: Double

    // MARK: - Init

    @objc public init(
        theme: SwrveButtonTheme,
        text: String?,
        frame: CGRect,
        calibration: SwrveCalibration?,
        renderScale: Double
    ) {

        self.theme = theme
        self.calibration = calibration
        self.renderScale = renderScale
        super.init(frame: frame)

        setTitle(text, for: .normal)
        applyTextAlignment()
        applyPadding()
        applyFont()
        applyFontColor()
        applyCornerRadius()
        applyBorder()
        applyBackground()

        clipsToBounds = true
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Highlight state

    @objc public override var isHighlighted: Bool {
        didSet {
            if isHighlighted {
                if let bg = theme.pressedState.bgColor {
                    backgroundColor = SwrveUtils.processHexColorValue(bg)
                }
                if let bc = theme.pressedState.borderColor {
                    layer.borderColor = SwrveUtils.processHexColorValue(bc).cgColor
                }
            } else {
                if let bg = theme.bgColor {
                    backgroundColor = SwrveUtils.processHexColorValue(bg)
                }
                if let bc = theme.borderColor {
                    layer.borderColor = SwrveUtils.processHexColorValue(bc).cgColor
                }
            }
        }
    }

    // MARK: - Apply helpers

    private func applyTextAlignment() {
        guard let align = theme.hAlign else { return }
        switch align {
        case "LEFT": contentHorizontalAlignment = .left
        case "CENTER": contentHorizontalAlignment = .center
        case "RIGHT": contentHorizontalAlignment = .right
        default: break
        }
    }

    private func applyFont() {
        let baseSize = CGFloat(truncating: theme.fontSize ?? 0)
        let defaultFont = UIFont.systemFont(ofSize: baseSize)
        let titleFont = SwrveSDKUtils.font(
            fromFile: theme.fontFile ?? "",
            postscriptName: theme.fontPostscriptName ?? "",
            size: baseSize,
            style: theme.fontNativeStyle ?? "",
            withFallback: defaultFont)

        var scaledPointSize = baseSize
        if let calibration {
            scaledPointSize = SwrveSDKUtils.scale(
                titleFont,
                calibration: calibration,
                swrveFontSize: baseSize,
                renderScale: CGFloat(renderScale))
        }
        titleLabel?.font = titleFont.withSize(scaledPointSize)

        if theme.truncate {
            titleLabel?.lineBreakMode = .byTruncatingTail
        } else {
            titleLabel?.numberOfLines = 1
            titleLabel?.lineBreakMode = .byWordWrapping
            var log = "SwrveThemedUIButton initial size: \(titleLabel?.font.pointSize ?? 0)"
            SwrveLogger.logDebug(log)
            scaleDownText(beginSize: scaledPointSize)
            log = "SwrveThemedUIButton after scale down size: \(titleLabel?.font.pointSize ?? 0)"
            SwrveLogger.logDebug(log)
            titleLabel?.adjustsFontSizeToFitWidth = true
        }
    }

    private func scaleDownText(beginSize: CGFloat) {
        guard let text = titleLabel?.text, !text.isEmpty, !bounds.size.equalTo(.zero) else { return }

        let left = CGFloat(truncating: theme.leftPadding ?? 0)
        let right = CGFloat(truncating: theme.rightPadding ?? 0)
        let top = CGFloat(truncating: theme.topPadding ?? 0)
        let bottom = CGFloat(truncating: theme.bottomPadding ?? 0)
        let border = CGFloat(truncating: theme.borderWidth ?? 0)

        let widthPaddingOffset = (left + right + border) * renderScale
        let heightPaddingOffset = (top + bottom + border) * renderScale

        let width = frame.size.width - widthPaddingOffset
        let height = frame.size.height - heightPaddingOffset
        guard width > 0, height > 0 else { return }

        let constraint = CGSize(width: width, height: height)
        var size = beginSize

        while size > 1 {
            if let currentFont = titleLabel?.font {
                titleLabel?.font = currentFont.withSize(size)
            }
            let textRect = titleLabel?.sizeThatFits(constraint) ?? .zero
            if ceil(textRect.height) <= height && ceil(textRect.width) <= width {
                break
            }
            size -= 1.0
        }
    }

    private func applyFontColor() {
        if let fc = theme.fontColor {
            setTitleColor(SwrveUtils.processHexColorValue(fc), for: .normal)
        }
        if let pc = theme.pressedState.fontColor {
            setTitleColor(SwrveUtils.processHexColorValue(pc), for: .highlighted)
        }
        if let focused = theme.focusedState?.fontColor {
            if #available(iOS 9.0, tvOS 9.0, *) {
                setTitleColor(SwrveUtils.processHexColorValue(focused), for: .focused)
            }
        }
    }

    private func applyCornerRadius() {
        var radius = CGFloat(truncating: theme.cornerRadius ?? 0) * renderScale
        radius = min(radius, frame.size.height / 2.0, frame.size.width / 2.0)
        layer.cornerRadius = radius
    }

    private func applyBorder() {
        let bw = CGFloat(truncating: theme.borderWidth ?? 0)
        guard bw > 0 else { return }
        if let borderColor = theme.borderColor {
            layer.borderColor = SwrveUtils.processHexColorValue(borderColor).cgColor
        }
        layer.borderWidth = bw * CGFloat(renderScale)
    }

    private func applyBackground() {
        if let bg = theme.bgColor, !bg.isEmpty {
            backgroundColor = SwrveUtils.processHexColorValue(bg)
        } else {
            setBackgroundImage(createUIImage(assetName: theme.bgImage), for: .normal)
            setBackgroundImage(createUIImage(assetName: theme.pressedState.bgImage), for: .highlighted)
            if let _ = theme.focusedState?.bgImage, #available(iOS 9.0, tvOS 9.0, *) {
                setBackgroundImage(createUIImage(assetName: theme.focusedState?.bgImage), for: .focused)
            }
        }
    }

    private func createUIImage(assetName: String?) -> UIImage? {
        guard let name = assetName, !name.isEmpty else { return nil }
        let cacheFolder = SwrveLocalStorage.swrveCacheFolder()
        let target = (cacheFolder! as NSString).appendingPathComponent(name)
        let url = URL(fileURLWithPath: target)
        guard let data = try? Data(contentsOf: url) else { return nil }
        return UIImage(data: data)
    }

    private func applyPadding() {
        let bw = CGFloat(truncating: theme.borderWidth ?? 0)
        let scaleValue = CGFloat(renderScale)
        let top = (CGFloat(truncating: theme.topPadding ?? 0) + bw) * scaleValue
        let left = (CGFloat(truncating: theme.leftPadding ?? 0) + bw) * scaleValue
        let bottom = (CGFloat(truncating: theme.bottomPadding ?? 0) + bw) * scaleValue
        let right = (CGFloat(truncating: theme.rightPadding ?? 0) + bw) * scaleValue
        contentEdgeInsets = UIEdgeInsets(top: top, left: left, bottom: bottom, right: right)
    }
}
